packages/playground/data-liberation/src/import/WP_Topological_Sorter.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+/**
+ * The topological sorter class.
+ *
+ * We create an in-memory index that contains offsets and lengths of items in the WXR.
+ * The indexer will also topologically sort posts so that the order we iterate over posts
+ * ensures we always get parents before their children.
+ */
+class WP_Topological_Sorter {
```

Comment:
In your experience, how well-suited is creating an in-memory index for, say, 1TB XML file?

Also, there's one catch – this will need to be re-entrant. We may run the import script 180 times before every bit of content is inserted into the database and ideally we won't have to recompute that index 180 times. It may also happen, that we'll pause before the index is complete.

Reply:
> We do not need to sort attachments here. They have already been parsed.

What if all the attachments for a post come before that post?

Reply:
> What if all the attachments for a post are listed before that post?

I now use a different step, so we must sort them out, but it's okay. I am creating tests for the worst-case scenarios.

Reply:
> In your experience, how well-suited is creating an in-memory index for, say, 1TB XML file?

Considering the preexisting post found in a Playground standard installation, which is `6,800 bytes`:

```
Page XML bytes: 6,800 bytes
1TB / 6,800 bytes = ~147,000 pages
147,000 * 8 bytes (size of an integer on 64-bit systems) = 1.176.000 bytes = 1.1 MB
```

So, 1.1 MB for 147k pages. The Zend engine will probably use more memory for arrays. I assume it will be 2 MB for every 100.000 pages.

PHP uses Quicksort, which performs poorly (O(n^2)) on already sorted arrays, but the posts will likely be sorted well. WordPress export sorted by ID. Let's consider using insertion sort if we get poor performance in the worst scenarios.